library/Freckle/App/Http/Cache/Memcached.hs

```diff
@@ -0,0 +1,144 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedRecordDot #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Freckle.App.Http.Cache.Memcached
+  ( memcachedHttpCacheSettings
+  , memcachedHttpCodec
+  , memcachedHttpCache
+  ) where
+
+import Freckle.App.Prelude
+
+import Blammo.Logging (MonadLogger, logWarnNS)
+import Codec.Serialise (Serialise (..), deserialiseOrFail, serialise)
+import qualified Data.ByteString.Lazy as BSL
+import Data.CaseInsensitive (CI)
+import qualified Data.CaseInsensitive as CI
+import Data.String (IsString (..))
+import Database.Memcache.Types (Value)
+import Freckle.App.Http.Cache
+import Freckle.App.Memcached
+import qualified Freckle.App.Memcached.Client as Memcached
+import Freckle.App.OpenTelemetry (MonadTracer)
+import Network.HTTP.Client (Cookie, CookieJar, Request)
+import qualified Network.HTTP.Client.Internal as HTTP
+import Network.HTTP.Types.Header (ResponseHeaders)
+import Network.HTTP.Types.Status (Status)
+import Network.HTTP.Types.Version (HttpVersion (..))
+
+memcachedHttpCacheSettings
+  :: ( MonadUnliftIO m
+     , MonadLogger m
+     , MonadTracer m
+     , MonadReader env m
+     , HasMemcachedClient env
+     , Serialise body
+     )
+  => CacheTTL
+  -- ^ Default TTL, used when @max-age@ is not present
+  -> HttpCacheSettings m body Value
+memcachedHttpCacheSettings defaultTTL =
+  HttpCacheSettings
+    { shared = False
+    , cacheable = const True
+    , defaultTTL
+    , getCurrentTime = liftIO getCurrentTime
+    , logWarn = logWarnNS "http.cache" . fromString
+    , codec = memcachedHttpCodec
+    , cache = memcachedHttpCache
+    }
+
+memcachedHttpCodec :: Serialise body => HttpCacheCodec body Value
+memcachedHttpCodec =
+  HttpCacheCodec
+    { serialise = BSL.toStrict . serialise . fromResponse
+    , deserialise = \req ->
+        bimap show (toResponse req)
+          . deserialiseOrFail
+          . BSL.fromStrict
+    }
```

Comment:
There is a subtle concern I want to think about carefully here.

So (nothing in this paragraph you don't already know but), a web service like assets.freckle is an API for which we have to think about backwards compatibility. We do this by not removing things unless we know they're not in use, by taking advantage of the fact that adding a new endpoint is a non-breaking changes, and by the fact that adding new fields to JSON objects is a non-breaking change.

The thing that I don't see talked about as much: A cache layer on top of such a service is *another* API that also has to preserve backwards compatibility. It's not as big a deal because things only stay in the cache for five minutes, but it still means that a breaking change to a `Serialise` instance could easily produce a five-minute outage when we deploy a new version of fancy-api that expects a different serialization format from what's in memcache. Or worse, maybe we get unlucky enough that deserialization actually succeeds and ends up producing erroneous results.

The nice property of JSON that adding new fields is a non-breaking change is - I doubt? - something that holds for CBOR. So I think a system for versioning this API - to make sure that the cache keys change when the value format changes - is really worth considering.

This is a big part of why I'm coming around to the idea of caching exactly the server's response rather than transforming it first, because whatever that transformation is, it's something that also needs to be versioned. Which gets really really hard when you're dealing with nested records of types with `Serialise` instances, because, just like with JSON, a change to a type in one place can easily end up having an action at a distance on some API that you didn't know was versioned.

I'm not sure I've thought this through all the way, but my thought is that `Serialise` is a footgun here that we should avoid in favor of something value-level codec type which represents a `Serialise` instance plus a version identifier, like

```haskell
data Codec a = Codec
  { encode :: a -> Encoding
  , decode :: Decoder s a
  , version :: Tree Text
  }
```

```haskell
memcachedHttpCodec :: Codec body -> HttpCacheCodec body Value
```

... `version` being a `Tree` where the root node is the version of your codec, and the children are version identifiers from the codecs that you use. Probably a hash of that tree then ends up in the cache keys.

I was going for something similar with [cacheKeyContribution](https://github.com/freckle/content-client/blob/a840a4f8b1a531f7f9f6e6fd820f468f89d91984/content-client-core/library/Freckle/Content/Client/Core/Abstract/Exchange.hs#L65) in content-client - The idea being that when you change types, you should increment that value to result in moving to new cache keys - but I don't think what I did there was robust enough to solve the problem, because I'm still using a lot of generic-derived `Serialise` instances and I don't imagine that anyone modifying types in the library will remember to go looking for versions to bump according to the resulting cascading changes to the CBOR API that we never see or think about.

Reply:
> I guess I didn't mean raw bytes on the wire, but rather that it's easier to get right if there is a single mapping from http-client's response type to our cached format rather than bespoke formats for each route that experience a lot of minor shifts all the time.

Are you just suggesting we use `Response ByteString` here instead of `SomeConstraint body => Reponse body`?

I'm fine with that, it's simpler and is how it worked in an earlier version. I thought you described caching a more compact format, post-JSON processing as somewhat of a goal, so I was reaching for it here.

> It is, that's the point

Gotcha. Yeah, so what I'm saying is you could take on all that complexity to implement such a thing, or you could just lean on the "de-serialization failures behave like a cache miss" that we have today and the behavior is the same. Better even, since the latter will still hit when it can. The former requires developers remember to take an extra action, which they'll forget, and it will _always_ generate misses, whether there actually is a format change that impacted your cached values or not.

Reply:
> I thought you described caching a more compact format, post-JSON processing as somewhat of a goal

I did, but it seems less desirable as a wild west of myriad datatypes with unversioned binary formats.

Reply:
> I'd like to understand more about what kind of breaking changes occur in Serialise and how likely valid-but-different really is though, before I jump off that cliff.

Any time you change the order of fields of the same type in a record you're going to have this hazard, for one thing.

```haskell
data Person_v1 = Person_v1 { name :: Text, password :: Text }
  deriving stock (Generic, Show) deriving anyclass Serialise
data Person_v2 = Person_v2 { password :: Text, name :: Text }
  deriving stock (Generic, Show) deriving anyclass Serialise
```

```haskell
ghci> cachedValue = serialise Person_v1{ name="Chris", password="xyz" }
ghci> deserialise @Person_v2 cachedValue
Person_v2 {password = "Chris", name = "xyz"}
```

Reply:
OK, yeah that's 100% a deal-breaker for anything other than `Response ByteString` as our target. Done.

Reply:
FWIW, I was talking about upgrading `serialise` the library not actually run-of-the-mill changes to your own datatypes. That is why I didn't realize that just backing off to storing `Response ByteString` was a fix, since library updates still impact that.

Changing your own types would be waaaaaay more common situation, which (now that I see what you mean) is not something we could ignore.

Reply:
I think we talked about this before but I don't recall exactly - Any transport compression on that ByteString has been undone at this phase, right? Maybe we should apply our own compression, store the responses gzipped or something? Could make a big difference on some of the larger JSON responses.

Reply:
fwiw `math/en/questions.json` is 63 meg raw, 5 meg gzipped.

Reply:
I don't think that's handled at all at this level. However you would make your request and interpret the returned `ByteString` today is how it would work with this caching in place.

For example, if you made a request with `Accept: gzip` with `Network.HTTP.Client`, I'm relatively certain you need to decompression that response body you get back. That would be unchanged by putting this cache in between.

I agree it could be valuable to automatically add that `Accept` and decompress (based on the returned `Content-Type`) so that our cached values are smaller. Maybe we explore adding that later / next?

Reply:
> if you made a request with Accept: gzip with Network.HTTP.Client, I'm relatively certain you need to decompress that response body you get back yourself

I'm skeptical because [http-client](https://hackage.haskell.org/package/http-client-0.7.16/docs/Network-HTTP-Client.html#v:requestHeaders) says

> the Accept-Encoding header is set implicitly to gzip for convenience by default

Reply:
I'm going to invent a Semigroup, aren't I?

```hs
instance MonadHttp (HandlerFor App) where
  httpLbs = httpGzip
    <> httpCached memcachedHttpCacheSettings
    <> httpRetry429s
    <> maybe httpStubbed (hoistIO httpLbs) appStubs
```

Reply:
Oh, how about that.

Heh, well now I'm back to -- do we need to rewrite `Network.HTTP.Client` to accomplish that feature?

Reply:
Set `decompress = const False` on the `Request` to stop it from decompressing automatically? Then... maybe there's some stuff in `Network.HTTP.Client.Internal` that can be used but yeah I guess you'd be left with rewriting some of the unzipping logic there.